torch/distributed/checkpoint/_hf_storage.py

```diff
@@ -0,0 +1,201 @@
+        return plans
+
+    def write_data(
+        self,
+        plan: SavePlan,
+        planner: SavePlanner,
+    ) -> Future[List[WriteResult]]:
+        # storage_plan is a map from key to file index
+        storage_plan: Dict[str, int] = plan.storage_data
+
+        buckets = self._split_by_storage_plan(storage_plan, plan.items)
+        highest_index = max(buckets.keys())
+
+        file_queue: queue.Queue = queue.Queue()
+        for file_index, write_items in buckets.items():
+            file_name = self._gen_file_name(file_index, highest_index)
+            file_queue.put(
+                (self.fs.concat_path(self.path, file_name), file_name, write_items)
+            )
+
+        return super()._write_data(planner, file_queue, safe_tensors=True)
+
+    def finish(self, metadata: Metadata, results: list[list[WriteResult]]) -> None:
+        metadata_to_write = {}
+        storage_md = {}
+        total_size = 0
+        for wr_list in results:
+            storage_md.update(
+                {wr.index.fqn: wr.storage_data.relative_path for wr in wr_list}
+            )
+            total_size += sum([wr.storage_data.length for wr in wr_list])
+        metadata_to_write["metadata"] = {"total_size": total_size}
+        metadata_to_write["weight_map"] = storage_md
+
+        metadata_path = self.fs.concat_path(self.path, f"{_metadata_fn}")
+        with self.fs.create_stream(metadata_path, "w") as metadata_file:
+            json.dump(metadata_to_write, metadata_file, indent=2)
+
+    def _split_by_storage_plan(
+        self, storage_plan: Dict[str, int], items: List[WriteItem]
+    ) -> Dict[int, List[WriteItem]]:
+        # storage_plan is a map from key to index
+        buckets = {}
+        for item in items:
+            key = item.index.fqn
+            idx = storage_plan[key]
+            if idx not in buckets:
+                buckets[idx] = [item]
+            else:
+                buckets[idx].append(item)
+
+        return buckets
+
+    def _gen_file_name(self, index: int, largest_index: int) -> str:
+        return (
+            FILE_NAME.format(
+                cpt_idx=f"{index}".zfill(5), num_shards=f"{largest_index}".zfill(5)
+            )
+            + SUFFIX
+        )
+
+    @property
+    def metadata_path(self) -> str:
+        return _metadata_fn
+
+
+class _HuggingFaceStorageReader(FsspecReader):
+    """
+    A reader that reads from a huggingface repository in the huggingface format.
+    Uses in Fsspec back-end to communicate with the huggingface hub.
+    """
+
+    def __init__(self, path: str, token: Optional[str] = None) -> None:
+        """
+        Initialize the huggingface reader pointing to path.
+
+        Args:
+            path: hf directory where the checkpoint will be read from. Should begin with hf://.
+            token: The token to use to authenticate with huggingface hub.
+        """
+        import fsspec  # type: ignore[import-untyped]
```

Comment:
Like wise, any reason for the lazy import? Is it really not imported by FsspecReader?